### ISSUE TITLE
model(img, size=640) doesn't work the same that --img 640 works! 

### ISSUE THREAD

**Issue by JonasMArnold** [ultralytics/yolov5#11403]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and found no similar bug report.

### YOLOv5 Component

Detection

### Bug

When I'm running a detection task via `cmd` (argparse): `python detect.py --weights best.pt --img 640 --conf 0.4` the appropriate result is being yielded:

> image 1/1 C:\frame0001.jpg: 384x640 10 entities, 148.7ms
> Speed: 0.0ms pre-process, 148.7ms inference, 0.0ms NMS per image at shape (1, 3, 640, 640)

However, if I do the same via the `size` argument, nothing happens (and the results are less accurate):

```
model = torch.hub.load('.', 'custom', path='C:/best.pt', source='local')
model.conf = 0.4
model.eval()

image = cv2.imread(image_path)
image = cv2.resize(image, (640, 384))

detections = model(image, size=640)
```

The result from using` size=640` (off-topic: I also used `size=(640, 640)`!): 

> image 1/1: 384x640 10 entities
> Speed: 0.0ms pre-process, 187.5ms inference, 0.0ms NMS per image at shape (1, 3, 384, 640)

As can clearly be seen, the 384, 640 size wasn't changed the same way that the `argparse `argument changed the processing.

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by JonasMArnold** [comment]
To re-phrase: I'm not stating that the `size` parameter doesn't work (in contrast to `--img 640`). It just doesn't resize the rectangular picture properly to the shape (e.g. 640, 640) even though specified.
It keeps up the ratio which shouldn't happen if it's also the equivalent in the cmd version.

**Comment by glenn-jocher** [comment]
Thank you for clarifying your issue, @JonasMArnold. So, to ensure I understand your request, can you please confirm that when you use the `size` parameter to perform detection, the rectangular image you are working with isn't being resized correctly to the specified dimensions (e.g. 640x640), and maintains the same aspect ratio as the original rectangular image? You are requesting that the image is resized to the specified dimensions, regardless of the original aspect ratio of the image, in the same manner as when you use the `--img 640` command line argument.

**Comment by JonasMArnold** [comment]
That is indeed correct @glenn-jocher !

The command line argument `--img 640` resizes it to 640x640, whereas `size = 640` (or `size = (640, 640)` for that matter) resizes it to the rectangular ratio (same aspect ratio) and thus to 384x640. In this process, annotations differ and are less accurate sometimes.

**Comment by glenn-jocher** [comment]
Thank you for confirming, @JonasMArnold. Please understand that `size` parameter resizes the image to the closest dimensions to `size` that maintain its aspect ratio. In contrast, the `--img 640` specifies an exact dimension to resize the image to, which may lead to some distortion if the original aspect ratio is different from 640x640. To achieve the same behavior with the `size` parameter, you'll need to add some padding to the image to match the specified dimensions while maintaining its aspect ratio. There are many ways to accomplish this, including using OpenCV to add the necessary padding. Please let me know if this information helps, or if there is anything else I can do for you.

**Comment by JonasMArnold** [comment]
What you described is the observed behavior, @glenn-jocher - but is this wanted?
Considering you can also parse in a tuple (e.g. (640, 640), but only the long-side is used to calculate the aspect ratio...)? One would assume you could force the padding to be added if you're able to parse in the size. Am I missing something here?

**Comment by glenn-jocher** [comment]
You are correct, @JonasMArnold, and I apologize for any confusion my earlier response caused. To clarify, when using the `size` parameter, the image is resized to the largest possible dimensions (while maintaining the aspect ratio) that are less than or equal to the values specified in the `size` parameter. However, this does not take into account where the longer image dimension lies. As a result, the value entered for the short dimension is disregarded (in your example, 384), and the image is resized to an aspect ratio that matches the specified long dimension (in your example, 640).

Unfortunately, this behavior cannot be modified without modifying the source code of YOLOv5. Therefore, as previously suggested, to achieve resizing that matches the behavior of `--img` command-line argument, you can either manually pad the image to the desired dimensions, or use other methods for resizing the image, such as `resize` and `copyMakeBorder` from OpenCV's `cv2` module. Let me know if you need any other help with this.

**Comment by JonasMArnold** [comment]
Shall I open a feature request to offer an additional parameter (or the functionality to check if a tuple was passed in and if so, re-size and pad the image respectively)? @glenn-jocher

**Comment by glenn-jocher** [comment]
Thank you for considering contributing to YOLOv5, @JonasMArnold. Yes, I think opening a feature request to add this functionality would be a good idea. It would be helpful to others in the community who might have similar issues with resizing images using the `size` parameter. Please note that there is already an open issue in the YOLOv5 Github repository regarding this matter, but you can open a new one if you think it specifies the problem you are describing more clearly. Additionally, in your feature request, it would be beneficial to explain the advantages of the proposed feature, how the feature would work, and provide code examples, if possible.
